plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy

```diff
@@ -199,65 +199,78 @@ class GoogleBatchTaskHandler extends TaskHandler implements FusionAwareTask {
+            )
 
-            if( task.config.getAccelerator().getType() )
-                accelerator.setType( task.config.getAccelerator().getType() )
+        allocationPolicy.putAllLabels( task.config.getResourceLabels() )
 
-            instancePolicy.addAccelerators(accelerator)
-            instancePolicyOrTemplate.setInstallGpuDrivers(true)
+        // use instance template if specified
+        final machineType = task.config.getMachineType()
+        if( machineType && machineType.startsWith('template://') ) {
+            instancePolicyOrTemplate
+                .setInstallGpuDrivers( executor.config.getInstallGpuDrivers() )
+                .setInstanceTemplate( machineType.minus('template://') )
         }
 
-        if( executor.config.cpuPlatform )
-            instancePolicy.setMinCpuPlatform( executor.config.cpuPlatform )
+        // otherwise create instance policy
+        else {
+            final instancePolicy = AllocationPolicy.InstancePolicy.newBuilder()
 
-        if( task.config.getMachineType() )
-            instancePolicy.setMachineType( task.config.getMachineType() )
+            if( task.config.getAccelerator() ) {
+                final accelerator = AllocationPolicy.Accelerator.newBuilder()
+                    .setCount( task.config.getAccelerator().getRequest() )
 
-        machineInfo = findBestMachineType(task.config)
-        if( machineInfo )
-            instancePolicy.setMachineType(machineInfo.type)
+                if( task.config.getAccelerator().getType() )
+                    accelerator.setType( task.config.getAccelerator().getType() )
 
-        if( executor.config.serviceAccountEmail )
-            allocationPolicy.setServiceAccount(
-                ServiceAccount.newBuilder()
-                    .setEmail(executor.config.serviceAccountEmail)
-            )
+                instancePolicy.addAccelerators(accelerator)
+                instancePolicyOrTemplate.setInstallGpuDrivers(true)
+            }
 
-        if( executor.config.preemptible )
-            instancePolicy.setProvisioningModel( AllocationPolicy.ProvisioningModel.PREEMPTIBLE )
+            if( executor.config.cpuPlatform )
+                instancePolicy.setMinCpuPlatform( executor.config.cpuPlatform )
 
-        if( executor.config.spot )
-            instancePolicy.setProvisioningModel( AllocationPolicy.ProvisioningModel.SPOT )
+            if( machineType )
+                instancePolicy.setMachineType( machineType )
 
-        // Fusion configuration
-        if( fusionEnabled() ) {
-            instancePolicy.addDisks(AllocationPolicy.AttachedDisk.newBuilder()
-                    .setNewDisk(AllocationPolicy.Disk.newBuilder()
-                            .setType("local-ssd")
-                            .setSizeGb(375)
-                    )
-                    .setDeviceName("fusion")
-            )
+            machineInfo = findBestMachineType(task.config)
+            if( machineInfo )
```

Comment:
@pditommaso @jordeu Should these two conditions be switched? Otherwise the custom machineType will be overwritten by the "best" machine type.

Reply:
Okay but this comment is about the `findBestMachineType()` method overwriting a custom machine type setting, I think the custom machine type should apply as long as it doesn't contain wildcards.

Reply:
If they only set one machine type `findBestMachineType` will always return that machine type. See here:
https://github.com/nextflow-io/nextflow/blob/1e38274ac573d8016c9704250b701b84e4ffcbbc/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L90-L97

I'm not checking for wildcards earlier in the code, because this logic also allows the user to define a list of machine types separated by comma, without any wildcards. On that case the `findBestMachineType` will select the best from the user provided list.

Reply:
Thanks Jordi, in that case we don't need to set `machineType` directly because `findBestMachineType()` will do that anyway.